### PR TITLE
SOC-785 Add Swagger Documentation to User Attribute Service

### PR DESCRIPTION
**Pull request by jsutterfield**
Ticket: https://wikia-inc.atlassian.net/browse/SOC-785

### CODE REVIEW

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java [hunk 1]
```diff
@@ -21,7 +22,11 @@
...
 
 
-// TODO update this to receive JSON (potentially pending based on what Platform prescribes).
+@Api(
+    basePath = "/",
+    value = "Attributes",
+    description = "Attributes available to users",
+    produces = RepresentationFactory.HAL_JSON)
 @Path("/attr")
 @Produces(RepresentationFactory.HAL_JSON)
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java]: Is this needed now that @Api contains this information?

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java [hunk 2]
```diff
@@ -30,7 +35,7 @@
 
   private URI createAttributeSelfUrl(String attribute) {
     return UriBuilder.fromResource(AttributeResource.class)
-        .path(AttributeResource.class, "getUserAttributeByName")
+            .path(AttributeResource.class, "getUserAttributeByName")
         .build(attribute);
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java]: The indentation of the previous two lines should match

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java [hunk 3]
```diff
@@ -59,17 +67,26 @@ public Response getAttributes() throws java.sql.SQLException {
...
-            .build())
-        .orElseGet(() -> StandardResponses
-            .NotFound(String.format("No attribute found with name %s", name)));
+      return optionalAttribute
+              .map(attribute -> Response.ok(UserAttributeApplication.REPRESENTATION_FACTORY
+                      .newRepresentation(createAttributeSelfUrl(name))
+                      .withBean(attribute))
+                      .build())
+              .orElseGet(() -> StandardResponses
+                      .NotFound(String.format("No attribute found with name %s", name)));
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java]: These look like they're indented two spaces too far.  Should line up w/ L82 no?

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 4]
```diff
@@ -49,13 +68,21 @@ public URI createUserSelfUrl(Integer userId) {
 
   public URI createUserAttributeSelfUrl(Integer userId, String attributeName) {
     return UriBuilder.fromResource(UserResource.class)
-        .path(UserResource.class, "getAttributeForUser")
+            .path(UserResource.class, "getAttributeForUser")
         .build(userId, attributeName);
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: Same as above, should line up, whether both 4 spaces or both 8.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 5]
```diff
@@ -49,13 +68,21 @@ public URI createUserSelfUrl(Integer userId) {
...
+      value = "Returns all available attributes for the specified userID",
+      response = UserAttribute.class
+  )
+  @ApiResponses(value = {
+     @ApiResponse(code = 404, message = "No attributes found for this userID")
+  })
   public Response getAllAttributesForUser(
-      @PathParam("userID") int userID
+     @ApiParam(value = "The ID of the user")
+     @PathParam("userID") int userID
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: More spacing weirdness above; seems like this should be 2 spaces or 4?

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 6]
```diff
@@ -64,15 +91,16 @@ public Response getAllAttributesForUser(
...
       representation.withRepresentation("properties",
                                         UserAttributeApplication.REPRESENTATION_FACTORY
-                                            .newRepresentation(createUserAttributeSelfUrl(userID,
-                                                                                          userAttribute
-                                                                                              .getName()))
-                                            .withBean(userAttribute)
+                                                .newRepresentation(
+                                                    createUserAttributeSelfUrl(userID,
+                                                                               userAttribute
+                                                                                   .getName()))
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: Seems like ```.getName```` should be attached to ```userAttribute``` on the same line.  Splitting on the dot makes sense for two or more chained methods, not as much with one.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 7]
```diff
@@ -125,7 +177,7 @@ public Response createAttributeForUser(
 
     userAttributeDAO.createAttributeForUser(userID, attrName, value);
     Representation representation = UserAttributeApplication.REPRESENTATION_FACTORY
-        .newRepresentation(createUserAttributeSelfUrl(userID, attrName))
+            .newRepresentation(createUserAttributeSelfUrl(userID, attrName))
         .withProperty("name", attrName)
         .withProperty("value", value);
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: Another alignment agreement issue.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 8]
```diff
@@ -150,7 +211,7 @@ public Response updateAttributeForUser(
 
     userAttributeDAO.updateAttributeForUser(userID, attrName, value);
     Representation representation = UserAttributeApplication.REPRESENTATION_FACTORY
-        .newRepresentation(createUserAttributeSelfUrl(userID, attrName))
+            .newRepresentation(createUserAttributeSelfUrl(userID, attrName))
```
Comment (garthwebb) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: And here

**Comment by garthwebb** [comment]
LGTM

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java [hunk 9]
```diff
@@ -29,11 +36,24 @@
...
+            type = "apiKey"),
+        @Authorization(
+            value = SwaggerBundleProvider.EXTERNAL_AUTH_NAME,
+            type = "apiKey")
+    })
 @Path("/user/{userID}")
 @Produces(RepresentationFactory.HAL_JSON)
 public class UserResource {
 
+  private final static String ATTRIBUTE_NOT_FOUND_MSG = "Attribute %s not found for user %d";
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/UserResource.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Move this constructor to comply with Java Code Conventions.](https://github.com/Wikia/pandora/blob/29a55a8a05dd14be1b86441f72b4b6f7abf65992/service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java#L46) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1213)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java [hunk 10]
```diff
@@ -2,21 +2,31 @@
...
+    implements DatabaseProvider<JooqConfigs>,
+               SwaggerConfigurationProvider {
 
   @NotNull
   @Valid
   @JsonProperty
   private JooqConfigs databases;
 
+  @JsonProperty
+  private SwaggerBundleConfiguration swaggerConfiguration;
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java]: MAJOR Remove this unused "swaggerConfiguration" private field. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1068)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Move this constructor to comply with Java Code Conventions.](https://github.com/Wikia/pandora/blob/aa1994251d20ddb865b6e52b15a36c935b1087a2/service/user-attribute/src/main/java/com/wikia/services/userattribute/resources/AttributeResource.java#L46) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1213)

**Pull request merged**